backend/python/tapa/task_graph.py

```diff
@@ -0,0 +1,285 @@
+    for arg in instance.args:
+      if arg.cat == Instance.Arg.Cat.ASYNC_MMAP:
+        v_name = rtl.async_mmap_instance_name(arg.mmap_name)
+        data_channel_width = max(port_name_to_width[arg.name], 32)
+        v_area = get_async_mmap_area(data_channel_width)
+        async_mmap_vertices[v_name] = {
+          'module': 'async_mmap',
+          'area': v_area
+        }
+
+  return type_marked(async_mmap_vertices, 'ASYNC_MMAP_VERTEX')
+
+
+def get_port_vertices(
+    top_task,
+    vitis_config_ini,
+):
+  """
+  FIXME: currently assume the DDR/HBM channels are 512 bit wide
+  """
+  port_vertices = {}
+
+  arg_name_to_external_port = parse_connectivity(vitis_config_ini)
+  for arg_name, external_port_name in arg_name_to_external_port.items():
+    port_cat, port_id = parse_port(external_port_name)
+    if port_cat == 'HBM':
+      port_vertices[arg_name] = {
+        'module': get_physical_port_vertex_name(arg_name),
+        'area': get_hbm_controller_area()
+      }
+    # the ddr will not overlap with user logic
+    elif port_cat == 'DDR':
+      port_vertices[arg_name] = {
+        'module': get_physical_port_vertex_name(arg_name),
+        'area': get_zero_area()
+      }
+    else:
+      raise NotImplementedError(f'unrecognized port type {port_cat}')
+
+  # check that every MMAP/ASYNC_MMAP port has a physical mapping
+  for arg_list in top_task.args.values():
+    for arg in arg_list:
+      if arg.cat in {Instance.Arg.Cat.ASYNC_MMAP, Instance.Arg.Cat.MMAP}:
+        if arg.name not in arg_name_to_external_port:
+          raise AssertionError(f'Missing physical binding for {arg.name} in {vitis_config_ini}')
+
+  return type_marked(port_vertices, 'PORT_VERTEX')
+
+
+def get_vertices(top_task: Task, vitis_config_ini: TextIO):
+  all_vertices = {}
+  all_vertices.update(get_task_vertices(top_task))
+  all_vertices.update(get_ctrl_vertices(top_task))
+  all_vertices.update(get_async_mmap_vertices(top_task))
+  all_vertices.update(get_port_vertices(top_task, vitis_config_ini))
+
+  return all_vertices
+
+
+#################################################################################
+
+
+def type_marked(name_to_obj, obj_type):
+  return {f'{obj_type}_{name}': obj for name, obj in name_to_obj.items()}
+
+def get_physical_port_vertex_name(arg_name):
+  return f'{arg_name}_external_controller'
+
+def get_axi_edge_name(port_arg):
+  port_vertex = get_physical_port_vertex_name(port_arg.name)
+  return f'axi_edge_from_{port_vertex}_to_{port_arg.instance.name}'
+
+def get_port_name_to_width(top_task):
+  """
+  FIXME: it is very confusing here. we are using "ports" to get the width table
+  but we are using "args" to look up the table
+  """
+  return {port.name: port.width for port in top_task.ports.values()}
+
+def make_autobridge_area(area: Dict[str, int]) -> Dict[str, int]:
+  return {{'BRAM_18K': 'BRAM'}.get(k, k): v for k, v in area.items()}
+
+
+def parse_connectivity(vitis_config_ini: TextIO) -> Dict[str, str]:
```

Comment:
Didn't this appear in `tapa.autobridge` already? Let's keep only one of them. Same with `parse_port`, I guess.

Reply:
Will discard tapa.autobridge later, refactor in progress